File: server_integration_tests/clients/java_reactive_client/client-app/src/test/java/dev/graphlink/server/SubscriptionsTest.java
package dev.graphlink.server;

import dev.graphlink.server.generated.client.GraphLinkClient;
import dev.graphlink.server.generated.inputs.ArticleCreatedFieldArgs;
import dev.graphlink.server.generated.inputs.ArticleUpdatedFieldArgs;
import dev.graphlink.server.generated.inputs.CreateArticleFieldArgs;
import dev.graphlink.server.generated.inputs.CreateArticleInput;
import dev.graphlink.server.generated.inputs.UpdateArticleFieldArgs;
import dev.graphlink.server.generated.inputs.UpdateArticleInput;
import dev.graphlink.server.generated.types.ArticleCreatedResponse;
import dev.graphlink.server.generated.types.ArticleUpdatedResponse;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class SubscriptionsTest {

    @Nested
    class ArticleCreated {
        @Test
        void emitsWhenAnArticleIsCreated() throws Exception {
            GraphLinkClient client = Fixtures.newClient();

            CompletableFuture<ArticleCreatedResponse> eventFuture =
                    client.subscriptions.articleCreated(new ArticleCreatedFieldArgs(2)).next().toFuture();

            Thread.sleep(300);

            var created = client.mutations.createArticle(
                    CreateArticleInput.builder().title("Subscribed Post").authorId("1").build(),
                    new CreateArticleFieldArgs(2)).block();

            var event = eventFuture.get(10, TimeUnit.SECONDS);
            assertEquals(created.getCreateArticle().getId(), event.getArticleCreated().getId());
            assertEquals("Subscribed Post", event.getArticleCreated().getTitle());
        }
    }

    @Nested
    class ArticleUpdated {
        @Test
        void emitsWhenTheMatchingArticleIsUpdated() throws Exception {
            GraphLinkClient client = Fixtures.newClient();

            var created = client.mutations.createArticle(
                    CreateArticleInput.builder().title("Will Update").authorId("2").build(),
                    new CreateArticleFieldArgs(2)).block();

            CompletableFuture<ArticleUpdatedResponse> eventFuture = client.subscriptions.articleUpdated(
                    created.getCreateArticle().getId(),
                    new ArticleUpdatedFieldArgs(2)).next().toFuture();

            Thread.sleep(300);

            var updated = client.mutations.updateArticle(
                    UpdateArticleInput.builder().id(created.getCreateArticle().getId()).title("Updated via subscription").build(),
                    new UpdateArticleFieldArgs(2)).block();

            var event = eventFuture.get(10, TimeUnit.SECONDS);
            assertEquals(updated.getUpdateArticle().getId(), event.getArticleUpdated().getId());
            assertEquals("Updated via subscription", event.getArticleUpdated().getTitle());
        }
    }
}

File: server_integration_tests/clients/kotlin_client/client-app/src/test/kotlin/dev/graphlink/server/SubscriptionsTest.kt
package dev.graphlink.server

import dev.graphlink.server.generated.inputs.ArticleCreatedFieldArgs
import dev.graphlink.server.generated.inputs.ArticleUpdatedFieldArgs
import dev.graphlink.server.generated.inputs.CreateArticleFieldArgs
import dev.graphlink.server.generated.inputs.CreateArticleInput
import dev.graphlink.server.generated.inputs.UpdateArticleFieldArgs
import dev.graphlink.server.generated.inputs.UpdateArticleInput
import kotlinx.coroutines.async
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout
import org.junit.jupiter.api.Nested
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.time.Duration.Companion.seconds

class SubscriptionsTest {

    @Nested
    inner class ArticleCreated {
        @Test
        fun emitsWhenAnArticleIsCreated() = runBlocking {
            val client = Fixtures.newClient()

            val eventDeferred = async {
                withTimeout(10.seconds) {
                    client.subscriptions.articleCreated(ArticleCreatedFieldArgs(2)).first()
                }
            }

            delay(300)

            val created = client.mutations.createArticle(
                CreateArticleInput(title = "Subscribed Post", authorId = "1"),
                CreateArticleFieldArgs(2),
            )

            val event = eventDeferred.await()
            assertEquals(created.createArticle.id, event.articleCreated.id)
            assertEquals("Subscribed Post", event.articleCreated.title)
        }
    }

    @Nested
    inner class ArticleUpdated {
        @Test
        fun emitsWhenTheMatchingArticleIsUpdated() = runBlocking {
            val client = Fixtures.newClient()

            val created = client.mutations.createArticle(
                CreateArticleInput(title = "Will Update", authorId = "2"),
                CreateArticleFieldArgs(2),
            )

            val eventDeferred = async {
                withTimeout(10.seconds) {
                    client.subscriptions
                        .articleUpdated(created.createArticle.id, ArticleUpdatedFieldArgs(2))
                        .first()
                }
            }

            delay(300)

            val updated = client.mutations.updateArticle(
                UpdateArticleInput(id = created.createArticle.id, title = "Updated via subscription"),
                UpdateArticleFieldArgs(2),
            )

            val event = eventDeferred.await()
            assertEquals(updated.updateArticle.id, event.articleUpdated.id)
            assertEquals("Updated via subscription", event.articleUpdated.title)
        }
    }
}
